### DISCUSSION TITLE
Trouble getting started

### DISCUSSION THREAD

**Discussion by dmb0058** [clientIO/joint#2541] [Q&A] [answered]
### Introduction

Sounds like it might be a FAQ, but I can't find it asked before ... and I think this is a problem with my knowledge of how to use JS modules in the browser, but I'd be very grateful for any help to get going.

I've been following the JointJS tutorial, and have a graph with a few cells linked together, all good :)

Now I'd like to try layouts, but I'm finding it hard to get things working.

I have npm'd jointjs in the parent directory of my test app and am trying to import the DirectedGraph code. My closest attempt looks like this:

```
        <!-- content -->
	<div id="myholder"></div>

	<!-- dependencies -->
	<script type="text/javascript" src="../node_modules/@joint/core/dist/joint.js"></script>

	<!-- code -->
	<script type="module">
		import { DirectedGraph } from "../node_modules/@joint/layout-directed-graph/dist/DirectedGraph.js";

		var namespace = joint.shapes;
		var graph = new joint.dia.Graph({}, { cellNamespace: namespace });

etc. etc.
```
which gives me this error in the console:
```
helloworld.html:24 Uncaught SyntaxError: The requested module '../node_modules/@joint/layout-directed-graph/dist/DirectedGraph.js' does not provide an export named 'DirectedGraph' (at helloworld.html:24:12)
```

I've also tried to use "'../node_modules/@joint/layout-directed-graph/DirectedGraph.mjs", and I tried to install the jointjs git code to see if I could get the directed graph layout example working but ran into trouble there too.

Thanks for any ideas!

David

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by zbynekstara** [comment]
Hi David,

In your case, the DirectedGraph code should be added as a dependency rather than as a module, as shown in the code snippet below (note that we also need to add dagre and graphlib, dependencies of DirectedGraph, in the same way):
```
<!-- dependencies -->
<script src="node_modules/@joint/core/dist/joint.js"></script>
<script src="node_modules/@joint/layout-directed-graph/dist/DirectedGraph.js"></script>
<script src="node_modules/@dagrejs/dagre/dist/dagre.js"></script>
<script src="node_modules/@dagrejs/graphlib/dist/graphlib.js"></script>
```

When added in this way, DirectedGraph becomes accessible via the `joint.layout` namespace. For example:
```
const graphBBox = joint.layout.DirectedGraph.layout(graph, {
    nodeSep: 50,
    edgeSep: 80,
    rankDir: "TB"
});
```

The documentation does not make this clear, so that is something that we will have to improve on our end.

Best regards,
Zbynek

**Comment by dmb0058** [comment]
Thanks Zbynek,

I think that takes me closer to the solution, as I hadn’t included dagrejs. I’ve added the lines below, but get this error in the console now, so not quite there yet:

DirectedGraph.js:265 Uncaught TypeError: graphlibUtil__namespace.Graph is not a constructor
    at Object.toGraphLib (DirectedGraph.js:265:27)
    at Object.layout (DirectedGraph.js:150:41)
    at helloworld.html:128:48

I’ll dig into it but any advice welcome!

David

**Reply by kumilingus** [reply, marked as answer]
I'd say that order of dependencies matters here.

```html
<script src="node_modules/@joint/core/dist/joint.js"></script>
<script src="node_modules/@dagrejs/graphlib/dist/graphlib.js"></script>
<script src="node_modules/@dagrejs/dagre/dist/dagre.js"></script>
<script src="node_modules/@joint/layout-directed-graph/dist/DirectedGraph.js"></script>
```

**Reply by dmb0058** [reply]
Got it - thanks so much to both of you!
